optimum/habana/transformers/models/llava/modeling_llava.py

```diff
@@ -0,0 +1,303 @@
+                    _pad_inputs(input_ids,
+                            attention_mask,
+                            self.config.image_token_index,
+                            self.vision_tower.vision_model.embeddings.num_patches,
+                            self.pad_token_id,
+                            vision_feature_select_strategy=self.config.vision_feature_select_strategy)
+
+        past_length = 0
+        if past_key_values is not None:
+            if token_idx is None:
+                if isinstance(past_key_values, Cache):
+                    cache_length = past_key_values.get_seq_length()
+                    past_length = past_key_values.seen_tokens
+                    max_cache_length = past_key_values.get_max_length()
+                else:
+                    cache_length = past_length = past_key_values[0][0].shape[2]
+                    max_cache_length = None
+                # Keep only the unprocessed tokens:
+                # 1 - If the length of the attention_mask exceeds the length of input_ids, then we are in a setting where
+                # some of the inputs are exclusively passed as part of the cache (e.g. when passing input_embeds as
+                # input)
+                if attention_mask is not None and attention_mask.shape[1] > input_ids.shape[1]:
+                    input_ids = input_ids[:, -(attention_mask.shape[1] - past_length) :]
+                # 2 - If the past_length is smaller than input_ids', then input_ids holds all input tokens. We can discard
+                # input_ids based on the past_length.
+                elif past_length < input_ids.shape[1]:
+                    input_ids = input_ids[:, past_length:]
+                    # 3 - Otherwise (past_length >= input_ids.shape[1]), let's assume input_ids only has unprocessed tokens.
+                elif self.config.image_token_index in input_ids:
+                    input_ids = input_ids[:, input_ids.shape[1] - 1 :]
+                    # If the cache has seen more tokens than it can hold, then the cache has a size limit. Let's discard the
+                    # older attention values, as their corresponding values are not part of the input.
+                    if cache_length < past_length and attention_mask is not None:
+                        attention_mask = attention_mask[:, -(cache_length + input_ids.shape[1]) :]
+            else:
+                # past_length += token_idx
+                input_ids = torch.index_select(input_ids, 1, token_idx + image_offset - 1)
+
+        position_ids = kwargs.get("position_ids", None)
+        if attention_mask is not None and position_ids is None:
+            # create position_ids on the fly for batch generation
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            if past_key_values:
+                position_ids = torch.sum(attention_mask, dim=1).unsqueeze(-1) - 1
```

Comment:
It seems to be different here: https://github.com/huggingface/transformers/blob/v4.38.2/src/transformers/models/llava/modeling_llava.py#L544

Reply:
updated when `token_idx==None`